python/ray/data/datasource/dbapi2_datasource.py

```diff
@@ -0,0 +1,122 @@
+    return pa.Table.from_pylist(pylist)
+
+
+class DBAPI2Datasource(Datasource[ArrowRow]):
+    def __init__(self, connection_factory: Callable[[], Connection]):
+        self.connection_factory = connection_factory
+
+    def create_reader(self, sql: str) -> "Reader":
+        return DBAPI2Reader(sql, self.connection_factory)
+
+
+@contextmanager
+def connect(
+    connection_factory: Callable[[], Connection]
+) -> Generator[Connection, None, None]:
+    connection = connection_factory()
+    try:
+        yield connection
+    except:
+        # `rollback` is optional since not all databases provide transaction support.
+        if hasattr(connection, "rollback"):
+            connection.rollback()
+    finally:
+        connection.close()
+
+
+class DBAPI2Reader(Reader):
+
+    NUM_SAMPLE_ROWS = 100
+
+    def __init__(self, sql: str, connection_factory: Callable[[], Connection]):
+        self.sql = sql
+        self.connection_factory = connection_factory
+
+    @property
+    @lru_cache
+    def sample_block(self) -> Block:
+        with connect(self.connection_factory) as connection:
+            cursor = connection.cursor()
+            cursor.execute(
+                f"SELECT * FROM ({self.sql}) as T LIMIT {self.NUM_SAMPLE_ROWS}"
+            )
+            return cursor_to_block(cursor)
+
+    @property
+    @lru_cache
+    def num_rows(self) -> int:
+        with connect(self.connection_factory) as connection:
+            cursor = connection.cursor()
+            cursor.execute(f"SELECT COUNT(*) FROM ({self.sql}) as T")
+            return cursor.fetchone()[0]
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        if self.num_rows == 0:
+            return None
+
+        accessor = BlockAccessor.for_block(self.sample_block)
+        size_bytes_per_row = accessor.size_bytes() / accessor.num_rows()
+        return math.ceil(size_bytes_per_row * self.num_rows)
+
+    def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
+        if self.num_rows == 0:
+            return []
+
+        elif self.num_rows < parallelism:
+            parallelism = self.num_rows
+
+        num_rows_per_block = self.num_rows // parallelism
+        num_blocks_with_extra_row = self.num_rows % parallelism
+
+        sample_block_accessor = BlockAccessor.for_block(self.sample_block)
+        sample_block_size_bytes = sample_block_accessor.size_bytes()
+        sample_block_schema = sample_block_accessor.schema()
+
+        tasks = []
+        offset = 0
+        for i in range(parallelism):
+            num_rows = num_rows_per_block
+            if i < num_blocks_with_extra_row:
+                num_rows += 1
+
+            def create_read_fn(num_rows: int, offset: int):
+                def read_fn() -> Iterable[Block]:
+                    with connect(self.connection_factory) as connection:
+                        cursor = connection.cursor()
+                        cursor.execute(
```

Comment:
FYI this performance can be bad, so it essentially has cost of O(parallelism * cost_of_sql_query), e.g. in Databricks doc - https://docs.databricks.com/sql/language-manual/sql-ref-syntax-qry-select-offset.html

```
When paging through a result set using LIMIT and OFFSET the skipped rows still get processed.
These rows merely get suppressed from the result set. 
Pagination with this technique is not advised for resource-intensive queries.
```